### PR TITLE
vhost-user-blk device implementation

### PR DESCRIPTION
**Pull request by yangzhon**
@sameo @rbradford @sboeuf @chao-p @bjzhjing 

for #130 

Please help review this PR, which is only work in **SINGLE** queue, and multi-queue is still on working. 

Below is command for vhost-user-blk 
./target/debug/cloud-hypervisor \
        --cpus 2 \
        --memory size=1024M,file=/dev/hugepages \
        --kernel /home/chen/project/linux-stable/arch/x86/boot/compressed/vmlinux.bin \
        --cmdline "console=ttyS0 reboot=k panic=1 nomodules i8042.noaux i8042.nomux i8042.nopnp i8042.dumbkbd root=/dev/vda3" \
        --disk /home/chen/project/clear-18080-kvm.img  \
        --console off \
        --serial tty \
        --vhost-user-blk "sock=/var/tmp/vhost.1,num_queues=1,queue_size=128,wce=true"  \
        --rng

for issue #232 , i also did one patch to fix it , please @sboeuf review and try it, thanks.
https://github.com/intel/cloud-hypervisor/pull/173/commits/009394c8e85200430e37fa10e3f9c06efa51c355

### CODE REVIEW

File: vm-virtio/src/vhost_user/handler.rs [hunk 1]
```diff
@@ -7,7 +7,7 @@ use epoll;
 use vmm_sys_util::eventfd::EventFd;
 
 use super::net::CtlVirtqueue;
-use super::{VIRTIO_NET_CTRL_MQ_VQ_PAIRS_MAX, VIRTIO_NET_CTRL_MQ_VQ_PAIRS_MIN};
+use super::{VIRTIO_MQ_VQ_PAIRS_MAX, VIRTIO_MQ_VQ_PAIRS_MIN};
```
Comment (sboeuf) [inline vm-virtio/src/vhost_user/handler.rs]: Could you fix your commit message by replacing `gernal` with `generic`?

File: vm-virtio/src/vhost_user/blk.rs [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+        // Retrieve available features only when connecting the first time.
+        let mut avail_features = vhost_user_blk.get_features().map_err(Error::VhostUserGetFeatures)?;
+        // Let only ack features we expect, that is VIRTIO_F_VERSION_1.
+        if (avail_features & VIRTIO_F_VERSION_1_BITMASK) != VIRTIO_F_VERSION_1_BITMASK {
+            return Err(Error::InvalidFeatures);
+        }
+        avail_features =
+            VIRTIO_F_VERSION_1_BITMASK | VhostUserVirtioFeatures::PROTOCOL_FEATURES.bits();
+        vhost_user_blk
+            .set_features(avail_features)
```
Comment (sboeuf) [inline vm-virtio/src/vhost_user/blk.rs]: As mentioned on #69, I'd like to see this features negotiation being factorized so that both vu_fs, vu_net and vu_blk can use it.

File: vm-virtio/src/vhost_user/blk.rs [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+            vhost_user_blk: vhost_user_blk,
+            kill_evt,
+            avail_features,
+            acked_features: 0u64,
+            config_space,
+            queue_sizes: vec![queue_size; num_queues],
+        })
+    }
+
+    pub fn setup_vub(
```
Comment (sboeuf) [inline vm-virtio/src/vhost_user/blk.rs]: Same here, this is generic to vhost-user protocol and I would expect this to be factorized between all vhost-user implementations (fs, net, and blk).

File: vmm/src/config.rs [hunk 4]
```diff
@@ -56,14 +56,14 @@ pub enum Error<'a> {
...
-    /// Failed parsing vhost-user-net sock parameter.
-    ParseVuNetSockParam,
-    /// Failed parsing vhost-user-net queue number parameter.
-    ParseVuNetNumQueuesParam(std::num::ParseIntError),
-    /// Failed parsing vhost-user-net queue size parameter.
-    ParseVuNetQueueSizeParam(std::num::ParseIntError),
-    /// Failed parsing vhost-user-net vector number parameter.
-    ParseVuNetNumVectorsParam(std::num::ParseIntError),
+    /// Failed parsing vhost-user sock parameter.
+    ParseVuSockParam,
```
Comment (sboeuf) [inline vmm/src/config.rs]: Please replace `gernal` with `generic` in your commit message.

File: src/main.rs [hunk 5]
```diff
@@ -90,6 +90,18 @@ fn main() {
                 .help("Kernel command line")
                 .takes_value(true),
         )
+        .arg(
+            Arg::with_name("vublk")
```
Comment (sboeuf) [inline src/main.rs]: I would prefer `vu-blk` instead of `vublk`.

File: vm-virtio/src/block.rs [hunk 6]
```diff
@@ -584,7 +584,7 @@ impl<T: 'static + DiskFile + Send> VirtioDevice for Block<T> {
         self.acked_features |= v;
     }
 
-    fn read_config(&self, offset: u64, mut data: &mut [u8]) {
+    fn read_config(&mut self, offset: u64, mut data: &mut [u8]) {
```
Comment (sboeuf) [inline vm-virtio/src/block.rs]: I'm not sure about this change. The read config is really about the driver running in the guest asking about the virtio device what is its config. This means `read_config()` should never modify the device configuration.
I understand the configuration needs to be determined through the communication with the vhost-user daemon, but that means we get this information through the vhost-user socket.

**Comment by yangzhon** [comment]
> @yangzhon I have several comments regarding the factorization of the code, but this needs to be addressed in #69.
> 
> Could you please submit all the `vhost_rs` crate fixes in a separate PR? Those fixes are completely independent from `vhost-user-blk` implementation and we should merge them separately.

@sboeuf, Since vhost_rs has been merged into local cloud-hypervisor, i added those changes here. My previous plan is i will submit one another PR to  vhost_rs in Liujiang's github.  For those changes, you mean i need firstly send PR to vhost_rs side? and then add changes in cloud-hypervisor? thanks.

File: vm-virtio/src/block.rs (hunk 6 above)
Comment (yangzhon) [reply vm-virtio/src/block.rs]: @sboeuf , yes, but for user blk, there are some difference. The queue num and wce info are set in cloud-hypervisor cmdline, not from backend like SPDK, but those parameters are needed set into blk config, which are required in guest kernel.  For WCE, guest kernel can change this value, and set back to backend. So, read_config need mut setting here, thanks.

Comment (sboeuf) [reply vm-virtio/src/block.rs]: Not sure I understand. I was saying this function will be exclusively triggered when the guest will read the configuration value. And in case the guest tries to change the configuration, it should go through `write_config()` instead.
BTW, what's WCE?

**Comment by sboeuf** [comment]
@yangzhon 
>Since vhost_rs has been merged into local cloud-hypervisor, i added those changes here. My previous plan is i will submit one another PR to vhost_rs in Liujiang's github.

Oh sorry for the confusion, I meant to submit the `vhost_rs` changes through a separate PR on cloud-hypervisor. This way, this current PR is only about `vhost-user-blk`.
Makes sense?

**Comment by yangzhon** [comment]
> @yangzhon
> 
> > Since vhost_rs has been merged into local cloud-hypervisor, i added those changes here. My previous plan is i will submit one another PR to vhost_rs in Liujiang's github.
> 
> Oh sorry for the confusion, I meant to submit the `vhost_rs` changes through a separate PR on cloud-hypervisor. This way, this current PR is only about `vhost-user-blk`.
> Makes sense?

@sboeuf , make sense now, i will send another PR dedicated for vhost_rs in cloud-hypervisor firstly, thanks.

File: vm-virtio/src/block.rs (hunk 6 above)
Comment (yangzhon) [reply vm-virtio/src/block.rs]: WCE: Write Cache Enable, In guest kernel, "write back" and "write through" can be set to this parameter.

The user blk config info is coming from two parts, one part mostly from backend SPDK, another part from cloud-hypervisor cmdline(queue num and WCE). When we call read_config() through vhost_rs protocol to get SPDK's config info, we need add extra info into this config struct, like queue num and wce value. So, this read_config should be changed in cloud-hypervisor side.  This is difference with net or fs.

Below is user blk config info, which will be exposed to guest kernel
struct virtio_blk_config {
    capacity: u64,
    size_max: u32,
    seg_max: u32,
    cylinders: u16,
    heads: u8,
    sectors: u8,
    blk_size: u32,
    physical_block_exp: u8,
    alignment_offset: u8,
    min_io_size: u16,
    opt_io_size: u32,
    **wce: u8,**
    unused0: [u8; 1],
    **num_queues: u16,**
    max_discard_sectors: u32,
    max_discard_seg: u32,
    discard_sector_alignment: u32,
    max_write_zeroes_sectors: u32,
    max_write_zeroes_seg: u32,
    write_zeroes_may_unmap: u8,
    unused1: [u8; 3],
}

Comment (sboeuf) [reply vm-virtio/src/block.rs]: Ok so if I follow correctly your explanation, you will get `wce` and `num_queues` information from the cloud-hypervisor command line. That's exactly what's happening for `vhost-user-net` [here](https://github.com/intel/cloud-hypervisor/blob/master/vm-virtio/src/vhost_user/net.rs#L39) and more specifically [here](https://github.com/intel/cloud-hypervisor/blob/master/vm-virtio/src/vhost_user/net.rs#L88). I know `vhost-user-net` is not directly writing those information to `config_space` byte array, but that's basically what you need to do.
Look at how it's done for virtio-pmem for a good example [here](https://github.com/intel/cloud-hypervisor/blob/master/vm-virtio/src/pmem.rs#L298-L301).

Basically, when those information come from the cloud-hypervisor command line, you can set the virtio configuration as soon as you create the instance. And if needed, you can even store those information to have them handy at any point in time.

Comment (yangzhon) [reply vm-virtio/src/block.rs]: @sboeuf ,

The latest code remove mut from "fn read_config(& mut self, offset: u64, mut data: &mut [u8])", thanks.

File: vmm/src/config.rs [hunk 7]
```diff
@@ -71,6 +71,8 @@ pub enum Error<'a> {
     ParseVuQueueSizeParam(std::num::ParseIntError),
     /// Failed parsing vhost-user-net server parameter.
     ParseVuNetServerParam(std::num::ParseIntError),
+    /// Failed parsing vhost-user-blk config_wce parameter.
+    ParseVuBlkConfigWceParam(std::num::ParseIntError),
```
Comment (sboeuf) [inline vmm/src/config.rs]: `ParseVuBlkWceParam` would be better.

File: vmm/src/config.rs [hunk 8]
```diff
@@ -513,6 +516,63 @@ impl<'a> VhostUserNetConfig<'a> {
     }
 }
 
+pub struct VhostUserBlkConfig<'a> {
+    pub config_wce: u8,
```
Comment (sboeuf) [inline vmm/src/config.rs]: Only `wce` should be enough.

File: vmm/src/config.rs [hunk 9]
```diff
@@ -513,6 +516,63 @@ impl<'a> VhostUserNetConfig<'a> {
...
+        let mut config_wce_str: &str = "";
+
+        for param in params_list.iter() {
+            if param.starts_with("sock=") {
+                sock = &param[5..];
+            } else if param.starts_with("num_queues=") {
+                num_queues_str = &param[11..];
+            } else if param.starts_with("queue_size=") {
+                queue_size_str = &param[11..];
+            } else if param.starts_with("config_wce=") {
```
Comment (sboeuf) [inline vmm/src/config.rs]: `wce`

File: src/main.rs [hunk 10]
```diff
@@ -165,6 +165,18 @@ fn main() {
...
                 .min_values(1),
         )
+        .arg(
+            Arg::with_name("vhost-user-blk")
+                .long("vhost-user-blk")
+                .help(
+                    "Vhost user Block parameters \"sock=<socket_path>,\
+                     num_queues=<number_of_queues>,\
+                     queue_size=<size_of_each_queue>, \
+                     config_wce=<value_of_wce>\"",
```
Comment (sboeuf) [inline src/main.rs]: `wce`

File: vhost_rs/src/vhost_user/sock_ctrl_msg.rs [hunk 11]
```diff
@@ -176,10 +176,13 @@ fn raw_recvmsg(fd: RawFd, iovecs: &mut [iovec], in_fds: &mut [RawFd]) -> Result<
 
     // Safe because the msghdr was properly constructed from valid (or null) pointers of the
     // indicated length and we check the return value.
-    let total_read = unsafe { recvmsg(fd, &mut msg, libc::MSG_WAITALL) };
+    let total_read = unsafe { recvmsg(fd, &mut msg, 0) };
 
     if total_read == -1 {
-        return Err(Error::last());
+        let total_read = unsafe { recvmsg(fd, &mut msg, libc::MSG_WAITALL) };
```
Comment (sboeuf) [inline vhost_rs/src/vhost_user/sock_ctrl_msg.rs]: This is not an acceptable fix IMO. With this patch, virtio-fs will still fail randomly when not all data will be read at once. The reason is that `total_read` won't be `-1`, but it won't be the full message either.
I think we're trying to solve the problem the wrong way here. Using `MSG_WAITALL` should always work since it's waiting for the full message to be received.
Which vhost-user-blk backend are you using? It seems to me like something might be wrong from the backend, or that we are missing something else from the `vhost_rs` crate.

File: vm-virtio/src/vhost_user/blk.rs [hunk 12]
```diff
@@ -86,6 +86,11 @@ impl<'a> Blk {
             | 1 << VIRTIO_F_VERSION_1
             | VhostUserVirtioFeatures::PROTOCOL_FEATURES.bits();
 
+        if config_wce == 1 {
```
Comment (sboeuf) [inline vm-virtio/src/vhost_user/blk.rs]: Why `== 1`? Should not you check for `> 0` to cover any other value from `wce`?

File: vm-virtio/src/vhost_user/blk.rs [hunk 13]
```diff
@@ -214,8 +219,18 @@ impl VirtioDevice for Blk {
             error!("Failed to write config space");
             return;
         }
-        let (_, right) = self.config_space.split_at_mut(offset as usize);
-        right.copy_from_slice(&data[..]);
+        // In fact, write_config() only handle wce value in vhost-user-blk.
+        // so, we can only set wce value here.
+        // Since vhost_set_config callback function in SPDK is NULL, which show
+        // there is not any handle in SPDK side for set_config operation. The
+        // detailed info ref spdk_extern_vhost_pre_msg_handler() in SPDK.
```
Comment (sboeuf) [inline vm-virtio/src/vhost_user/blk.rs]: Could you please remove specific comments about SPDK?
Also, you're mentioning `wce` should be the only value being set from the guest, but with this current implementation, you're letting any config entry being updated from the guest.

Last thing, this commit should be squashed with the one introducing wce.

**Comment by yangzhon** [comment]
> This is not an acceptable fix IMO. With this patch, virtio-fs will still fail randomly when not all data will be read at once. The reason is that `total_read` won't be `-1`, but it won't be the full message either.
> I think we're trying to solve the problem the wrong way here. Using `MSG_WAITALL` should always work since it's waiting for the full message to be received.
> Which vhost-user-blk backend are you using? It seems to me like something might be wrong from the backend, or that we are missing something else from the `vhost_rs` crate.

The backend is SPDK.

The config message include 12bytes(message header)+12bytes(config header)+60bytes(virtio_blk length), when vhost-user-blk device get this 84 bytes info from SPDK with MSG_WAITALL setting in recvmsg, the recv operation will be stuck.  Let me continue to debug this. thanks.

File: vmm/src/config.rs (hunk 7 above)
Comment (yangzhon) [reply vmm/src/config.rs]: Yes, thanks

File: vm-virtio/src/vhost_user/blk.rs (hunk 12 above)
Comment (yangzhon) [reply vm-virtio/src/vhost_user/blk.rs]: wce is only 0 or 1, let me change it to bool type, thanks

File: vmm/src/config.rs (hunk 8 above)
Comment (yangzhon) [reply vmm/src/config.rs]: yes, wce is bool type now

File: vmm/src/config.rs (hunk 9 above)
Comment (yangzhon) [reply vmm/src/config.rs]: yes, wce only

File: src/main.rs (hunk 10 above)
Comment (yangzhon) [reply src/main.rs]: yes, wce=true|false now

File: vhost_rs/src/vhost_user/sock_ctrl_msg.rs (hunk 11 above)
Comment (yangzhon) [reply vhost_rs/src/vhost_user/sock_ctrl_msg.rs]: New patch to fix this, https://github.com/intel/cloud-hypervisor/pull/173/commits/009394c8e85200430e37fa10e3f9c06efa51c355

File: vm-virtio/src/vhost_user/blk.rs (hunk 12 above)
Comment (yangzhon) [reply vm-virtio/src/vhost_user/blk.rs]: change the wce to bool type now

File: vm-virtio/src/vhost_user/blk.rs (hunk 13 above)
Comment (yangzhon) [reply vm-virtio/src/vhost_user/blk.rs]: yes, remove SPDK related log, thanks

**Comment by yangzhon** [comment]
@sboeuf 

Please review this patch again, thanks.

As for test case, i am writing the vhost-user-blk rust backend like SPDK, and will add it as the test case for vhost-user-blk. thanks.

File: vhost_rs/src/vhost_user/master.rs [hunk 14]
```diff
@@ -361,7 +361,8 @@ impl VhostUserMaster for Master {
         // But what content should the payload contains for a get_config() request?
         // So current implementation doesn't conform to the spec.
         let hdr = node.send_request_with_config_body(MasterReq::GET_CONFIG, &body, None)?;
-        let (reply, buf, rfds) = node.recv_reply_with_payload::<VhostUserConfig>(&hdr)?;
+        let (reply, buf, rfds) =
+            node.recv_reply_with_payload::<VhostUserConfig>(body.size as usize, &hdr)?;
```
Comment (sboeuf) [inline vhost_rs/src/vhost_user/master.rs]: Even if this works, this is not solving the root of the issue. Please take a look at my branch https://github.com/sboeuf/cloud-hypervisor-1/tree/fix-vu-blk
I took your branch, I reverted two of your patches, and I fixed the `GET_CONFIG` command, which now means you shouldn't run into the issue you were seeing.

File: vm-virtio/src/vhost_user/blk.rs [hunk 15]
```diff
@@ -86,6 +86,11 @@ impl<'a> Blk {
             | 1 << VIRTIO_F_VERSION_1
             | VhostUserVirtioFeatures::PROTOCOL_FEATURES.bits();
 
+        if wce == true {
+            // SPDK should expose WCE feature by default.
```
[the remark this answers is not in the document]
Comment (yangzhon) [reply vm-virtio/src/vhost_user/blk.rs]: yes, will do it, thanks

File: vmm/src/config.rs [hunk 16]
```diff
@@ -550,6 +553,61 @@ impl<'a> VsockConfig<'a> {
...
+            } else if param.starts_with("queue_size=") {
+                queue_size_str = &param[11..];
+            } else if param.starts_with("wce=") {
+                wce_str = &param[4..];
+            }
+        }
+
+        let mut num_queues: usize = 1;
+        let mut queue_size: u16 = 128;
+        let mut wce: bool = true;
```
[the remark this answers is not in the document]
Comment (yangzhon) [reply vmm/src/config.rs]: wce value is default to true in Qemu side, i will do it in CLI help usage.

File: vhost_rs/src/vhost_user/master.rs (hunk 14 above)
Comment (yangzhon) [reply vhost_rs/src/vhost_user/master.rs]: I tried your branch, it is still failed in recvmsg for get_config info from backend.

as for the revert patch: **https://github.com/sboeuf/cloud-hypervisor-1/commit/e74e0cae02f500139b225795f4e35b5d521633c2**,  i also had another solution in vhost crate after my vhost_rs changes in cloud-hypervisor.https://github.com/jiangliu/vhost/commit/d5a42d1ea580dd46dcd6274bc67d2a5f970488c9

Comment (sboeuf) [reply vhost_rs/src/vhost_user/master.rs]: Yes I can see that this patch you're mentioning is very close to my patch. The difference is that I expect the caller to provide the array of bytes (the payload) instead of providing an empty array from the vhost_rs crate.
The only thing I realized I'm missing in my branch is this line https://github.com/jiangliu/vhost/commit/d5a42d1ea580dd46dcd6274bc67d2a5f970488c9#diff-70449da7a94b8102ba670b5029bf011bR559
Could you try by fixing this line?

Also, is the failure you're seeing the same you were seeing before?

Comment (sboeuf) [reply vhost_rs/src/vhost_user/master.rs]: Could you please try with my updated branch. I have added this line https://github.com/sboeuf/cloud-hypervisor-1/commit/927f16f017efa2b39c1f13f768c62a357b3b5166#diff-739b90af7d2cfe98f8eb767f56a0aad2R555 to my patch, making the buffer being the right size, which should prevent `readmsg()` from being stuck.

**Comment by yangzhon** [comment]
> @yangzhon
> Sorry I forgot that you need to add an integration test before we can merge the PR.
> Could you build the `vhost-user-blk` daemon from QEMU to validate very simple behavior that our cloud-hypervisor VM is booting when using `--vhost-user-blk` device? And additionally, validate you can read a file that you put inside the shared block image.

@sboeuf , thanks.

I am trying the PR you changed before with vhost-user-blk backend, but there are some issues in get_protocol features. Let me debug it and will do CI for this, thanks.

**Comment by yangzhon** [comment]
The qemu build command
1. ./configure --target-list=x86_64-softmmu
2. make vhost-user-blk

The backend start command
1. dd if=/dev/zero of=blk bs=1M count=64
2. ./vhost-user-blk -b /home/chen/path/blk -s /home/chen/path/vhost.socket

The rust vhost-user-blk command
./target/debug/cloud-hypervisor \
        --cpus 2 \
        --memory size=1024M,file=/dev/shm \
        --kernel /home/chen/project/linux-stable/arch/x86/boot/compressed/vmlinux.bin \
        --cmdline "console=ttyS0 reboot=k panic=1 nomodules i8042.noaux i8042.nomux i8042.nopnp i8042.dumbkbd root=/dev/vda3" \
        --disk /home/chen/project/clear-18080-kvm.img  \
        --console off \
        --serial tty \
        --vhost-user-blk "sock=/home/chen/path/vhost.socket,num_queues=1,queue_size=128,wce=true"  \
        --rng

How to check
1. enter into guest to find the virtio-blk device enumerated by guest kernel
# lsblk | grep -c 64M

**Comment by sboeuf** [comment]
@yangzhon 
Hi! So I fixed your test (and rebased on `master`), as you did forget to set `--net` to let the host access the guest through ssh. That's why `lsblk` command was not successful, as it was never run in the guest.

Also, I decided to go one step further and extend your test by creating a block image with an actual ext4 filesystem and a very simple file on it. This way we can validate that accessing the content of the file works.

File: src/main.rs [hunk 17]
```diff
@@ -1179,6 +1221,91 @@ mod tests {
...
+                    .ssh_command("lsblk | grep vdc | grep -c 16M")
+                    .unwrap_or_default()
+                    .trim()
+                    .parse::<u32>()
+                    .unwrap_or_default(),
+                1
+            );
+
+            // Mount the device
+            guest.ssh_command("mkdir mount_image").unwrap();
```
Comment (rbradford) [inline src/main.rs]: Mustn't call `.unwrap()` after an `aver_eq!`. This can lead to hard to debug SIGILL crashes (nested panics.)

File: src/main.rs [hunk 18]
```diff
@@ -1179,6 +1221,91 @@ mod tests {
...
+            guest
+                .ssh_command("sudo mount -t ext4 /dev/vdc mount_image/")
+                .unwrap();
```
Comment (rbradford) [inline src/main.rs]: Ditto.

File: src/main.rs [hunk 19]
```diff
@@ -1179,6 +1221,91 @@ mod tests {
...
+                tb,
+                guest
+                    .ssh_command("cat mount_image/foo")
+                    .unwrap_or_default()
+                    .trim(),
+                "bar"
+            );
+
+            // Unmount the device
+            guest.ssh_command("sudo umount /dev/vdc").unwrap();
```
Comment (rbradford) [inline src/main.rs]: Ditto.

File: src/main.rs [hunk 20]
```diff
@@ -1179,6 +1221,91 @@ mod tests {
...
+            guest.ssh_command("rm -r mount_image").unwrap();
```
Comment (rbradford) [inline src/main.rs]: Ditto.

File: vm-virtio/src/vhost_user/blk.rs [hunk 21]
```diff
@@ -0,0 +1,274 @@
...
+        unsafe { &(*(null() as *const $ty)).$field as *const _ as usize }
+    };
+}
+
+struct SlaveReqHandler {}
+impl VhostUserMasterReqHandler for SlaveReqHandler {}
+
+#[derive(Copy, Clone, Debug, Default)]
+#[repr(C, packed)]
+struct virtio_blk_config {
```
Comment (rbradford) [inline vm-virtio/src/vhost_user/blk.rs]: What's wrong with the version in virtio-bindings?

File: vm-virtio/src/vhost_user/blk.rs [hunk 22]
```diff
@@ -0,0 +1,274 @@
...
+        // so, we can only set wce value here.
+        if self.config_space[offset as usize] == data[0] {
+            return;
+        }
+        self.vhost_user_blk
+            .set_config(offset as u32, VhostUserConfigFlags::WRITABLE, data)
+            .expect("Failed to set config");
+        self.config_space[offset as usize] = data[0];
+    }
+
```
Comment (rbradford) [inline vm-virtio/src/vhost_user/blk.rs]: You need to implement `reset()` it's quite feasible that someone would want to boot from this.

Comment (sboeuf) [reply vm-virtio/src/vhost_user/blk.rs]: Oh yes that is a good point!

File: vm-virtio/src/vhost_user/blk.rs (hunk 21 above)
Comment (sboeuf) [reply vm-virtio/src/vhost_user/blk.rs]: https://github.com/intel/cloud-hypervisor/pull/292 addresses the issue, which means the virtio-bindings from 5.0 are up to date.

**Comment by sboeuf** [comment]
@rbradford 
I addressed all your comments, PTAL!

One thing to note is that I couldn't validate properly the `reset()` implementation works. The reason is the FW does not work with vhost-user-blk (because of the backend over which we have no control so far, since we're using some test binary from QEMU). And the FW does not work because it cannot find the EFI partition. Digging a bit, I found the FW fails here: https://github.com/intel/rust-hypervisor-firmware/blob/master/src/part.rs#L75
I assume the problem is due to the fact that FW only supports 512 blocks while the backend might expose the blocks with different size.

I still wanted to validate that booting from vhost-user-net was working and that's why I also added a new test for this.

**Comment by sboeuf** [comment]
@sameo PTAL

**Comment by sboeuf** [comment]
@rbradford @yangzhon FYI I had to patch the `vhost-user-blk` daemon on our QEMU branch. Look at this commit: https://github.com/sboeuf/qemu/commit/84d6230fc98734580d9e77b5ea10de345aa2e854. The reason is that we're storing our images (`os_disk` and `raw_os_disk`) as part of `/tmp`, and we mount `/tmp` as `tmpfs`. Problem is, `tmpfs` does not support files to be opened with `O_DIRECT` flag, which is exactly what the `vhost-user-blk` daemon is trying to do by default.

**Comment by yangzhon** [comment]
> @rbradford @yangzhon FYI I had to patch the `vhost-user-blk` daemon on our QEMU branch. Look at this commit: [sboeuf/qemu@84d6230](https://github.com/sboeuf/qemu/commit/84d6230fc98734580d9e77b5ea10de345aa2e854). The reason is that we're storing our images (`os_disk` and `raw_os_disk`) as part of `/tmp`, and we mount `/tmp` as `tmpfs`. Problem is, `tmpfs` does not support files to be opened with `O_DIRECT` flag, which is exactly what the `vhost-user-blk` daemon is trying to do by default.

@sboeuf , thanks for checking this issue. The backend from Qemu defaultly support "write through" mode,  and or you can change this mode to "write back" by command(echo "write back" > /sys/block/vdb/cache_type) in guest cmdline when we start VM.

**Comment by sboeuf** [comment]
@yangzhon 
>you can change this mode to "write back" by command(echo "write back" > /sys/block/vdb/cache_type) in guest cmdline when we start VM.

I understand but this would not work in this case. The VMM cannot even create the VM since the connection to the backend failed, because the backend terminated after it couldn't open the block image.
